lib/include/jxl/encode.h

```diff
@@ -669,47 +669,123 @@ JXL_EXPORT JxlEncoderStatus JxlEncoderAddImageFrame(
+ *
+ * - position: Represents the current position, in bytes, within the output
+ * stream where the encoded data will be written next. This position moves
+ * forward as data is written, and can also be adjusted through the optional
+ * seek callback, if provided. At this position the next write will occur.
  *
+ * - watermark:  A position in the output stream that ensures all bytes before
+ * this point are finalized and won't be changed by later writes.
+ *
+ * All fields but `seek` are required, `seek` is optional and can be NULL.
  */
-JXL_EXPORT JxlEncoderStatus
-JxlEncoderSetOutputCallback(JxlEncoderOutputCallback callback);
+struct JxlEncoderOutputProcessor {
+  /**
+   * Required.
+   * An opaque pointer that the client can use to store custom data.
+   * This data will be passed to the associated callback functions.
+   */
+  void* opaque;
+
+  /**
+   * Required.
+   * Acquires a buffer into which the library will write the output data.
+   *
+   * If the `size` argument points to 0 and the returned value is NULL, this
+   * will be interpreted as asking the output writing to stop. In such a case,
+   * the library will return an error. The client is expected to set the size of
+   * the returned buffer based on the suggested `size` when this function is
+   * called.
+   *
+   *
+   * @param opaque user supplied parameters to the callback
+   * @param size points to a suggested buffer size when called; must be set to
+   * the size of the returned buffer once the function returns.
+   * @return a pointer to the acquired buffer or NULL to indicate a stop
+   * condition.
+   */
+  void* (*get_buffer)(void* opaque, size_t* size);
+
+  /**
+   * Required.
+   * Notifies the user of library that the current buffer's data has been
+   * written and can be released. This function should advance the current
+   * position of the buffer by `written_bytes` number of bytes.
+   *
+   * @param opaque user supplied parameters to the callback
+   * @param written_bytes the number of bytes written to the buffer.
+   */
+  void (*release_buffer)(void* opaque, size_t written_bytes);
+
+  /**
+   * Optional, can be NULL
+   * Seeks to a specific position in the output. This function is optional and
+   * can be set to NULL if the output doesn't support seeking. Can only be done
+   * when there is no buffer. Cannot be used to seek before the watermark
+   * position.
+   *
+   * @param opaque user supplied parameters to the callback
+   * @param position the position to seek to, in bytes.
+   */
+  void (*seek)(void* opaque, uint64_t position);
+
+  /**
+   * Required.
+   * Sets a watermark on the output data, at a specific position.
+   * Seeking will never request a position before the watermark.
+   *
+   * @param opaque user supplied parameters to the callback
+   * @param watermark_position the position, in bytes, where the watermark
+   * should be set.
+   */
+  void (*set_watermark)(void* opaque, uint64_t watermark_position);
```

Comment:
'Watermark' is not a good name for this imo. Maybe something like committed_position would be clearer.

Reply:
This is of course not the meaning of watermark as in "watermark in a paper", but rather marking how high the water has risen.

Reply:
Yes, but that's not the most common meaning of that word so it's a bit confusing. Maybe something like 'gauge position' would be better to avoid that confusion, but imo it's even better to use a name that clarifies _what_ is being marked/gauged here. Something like 'committed position', 'confirmed position' or 'finalized position'.

Reply:
I also think that e.g. finalized_position would be a better name.

Reply:
renamed to `finalized_position` and similar accordingly